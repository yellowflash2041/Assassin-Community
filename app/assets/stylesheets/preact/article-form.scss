@import 'variables';

.articleformcontainer {
  padding-top: calc(80px + 0.2vw);
  @media screen and ( min-width: 500px ){
    padding-top: calc(98px + 0.2vw);
  }
  padding-bottom: calc(94px + 2vw);
  background: white;
  background: $light-gray;
}

.articleform {
  margin: 0 auto;
  width: calc(95% - 40px);
  max-width: 880px;
  padding: 0px 20px;
  box-shadow: 5px 5px 0px darken($light-medium-gray, 3%);
  border: 1px solid darken($light-medium-gray, 5%);
  border-radius: 3px;
  background: white;
  position:relative;
  .container{
    border: 1px solid $light-medium-gray;
  }
  .help-guide-title{
    padding-top: 20px;
    h1{
      background: $yellow;
    }
  }
}

.articleform__form {
  display: flex;
  flex-direction: column;
}

.articleform__errors{
  background: $red;
  color: white;
  font-weight: bold;
  margin-top: 12px;
  border-radius: 3px;
  padding: 15px;
}

.articleform__orgsettings{
  padding: 0px;
  padding-left: 0px;
  margin: 20px 2px 0px;
  font-size:17px;
  cursor: pointer;
  user-select: none;
  img{
    height: 28px;
    width: 28px;
    border-radius: 50px;
    vertical-align: -7px;
    margin-right: 4px;
  }
  button {
    background: transparent;
    border-radius: 3px;
    border: 0px;
    font-size: 0.8em;
    font-family: $helvetica-condensed;
    width: 80px;
    border-radius: 100px;
    padding: 1px 0px;
    margin-left: 10px;
    border: 3px solid white;
    &.no{
      background: $light-gray;
      color: $medium-gray;
    }
    &.yes{
      background: darken($green, 18%);
      color: white;
    }
  }
  @media screen and ( min-width: 600px ){
    font-size:23px;
    padding-left: 10px;
    img{
      height: 33px;
      width: 33px;
      vertical-align: -8px;
      margin-right: 8px;
    }
    button {
      width: 120px;
    }
  }
}

.articleform__mainimage{
  padding: 10px 0px;
  text-align: center;
  width: 100%;
  img {
    width: 100%;
    height: calc(42vw - 44px);
    background: #fcfcfc;
    @media screen and ( min-width: 985px ){
      height: 370px;
    }
    display: block;
    cursor: pointer;
    &:hover{
      opacity:0.8;
    }
  }
}

.articleform__title {
  height: 26px;
  width: 99%;
  font-size: 18px;
  word-break: break-word;
  border: 0;
  padding: 20px 0px 15px;
  font-weight: 600;
  @media screen and ( min-width: 600px ){
    font-size: 29px;
    height: 50px;
  }
}
.articleform__detailfields{
  display: flex;
  flex-direction: row;
  position: relative;
}
.articleform__tagswrapper{
  display: inline-block;
  flex-grow: 1;
  float: left;
  padding-right: 15px;
  @media screen and ( min-width: 500px ){
    padding-right: 0px;
  }
}

.articleform__tags {
  height: 40px;
  width: 100%;
  border-radius: 3px;
  padding: 8px 8px 7px;
  @media screen and ( min-width: 500px ){
    width: 98%;
  }
  @media screen and ( min-width: 600px ){
    height: 19px;
    border-radius: 100px;
    padding-left: 17px;
  }
  margin-left: -4px;
  resize: none;
  border: 0;
  font-size: 16px;
  font-family: $monospace;
  font-weight: bold;
  background: $light-gray;
}

.articleform__tagsoptions{
  position: absolute;
  left: 0;
  right: 0;
  top: 50px;
  @media screen and ( min-width: 600px ){
    top: 30px;
  }
  background: white;
  border: 1px solid $medium-gray;
  box-shadow: $bold-shadow;
  font-family: $monospace;
  z-index:20;
}
.articleform__tagoptionrow{
  padding: 10px;
  cursor:pointer;
  &:hover{
    background: lighten($green, 27%);
  }
}
.articleform__tagoptionrow--active{
  background: $green;
  &:hover{
    background: darken($green, 10%);
  }
}

.articleform__detailsButton{
  color: $black;
  background: $purple;
  border: 0px;
  border-radius: 300px;
  padding: 3px 18px;
  font-family: $helvetica-condensed;
  font-size: 14.4px;
  height: 34px;
  display: none;
  @media screen and ( min-width: 600px ){
    display: inline-block;
  }
  margin-left:25px;
  &:hover{
    background: lighten($green, 24%);
  }
  img{
    width: 21px;
    height: 21px;
    vertical-align: -4px;
    display: inline-block;
  }
  &.articleform__detailsButton--moreconfig {
    margin-left: 5px;
    background: $light-medium-gray;
    &.articleform__detailsButton--bottom{
      margin-left: 5px;
      width:50px;
    }
  }
  &.articleform__detailsButton--bottom{
    margin-left: -5px;
    margin-top:10px;
    margin-bottom: 8px;
    display: inline-block;
    padding: 2px 15px;
    border-radius: 3px;
    width:100px;
    @media screen and ( min-width: 600px ){
      display: none;
    }  
  }
}

.articleform__description {
  height: 30px;
}

.articleform__body {
  min-height: calc(90vh - 370px);
  // height: calc(90vh - 240px);
  width: 98%;
  border: 0;
  font-size: 15px;
  resize: none;
  padding: 20px 6px;
  font-family: $monospace;
  @media screen and ( min-width: 600px ){
    font-size: 18px;
    min-height: calc(90vh - 280px);
  }
}

.articleform__buttons{
  padding: 20px 0px 20px;
  text-align: center;
  position:fixed;
  bottom: 0px;
  left: 0;
  right: 0;
  background: white;
  z-index: 20;
  border-top: 1px solid $light-medium-gray;
  button, a{
    cursor:pointer;
    background: $light-gray;
    font-family: $helvetica-condensed;
    font-size:14.5px;
    border: 0px;
    border-radius: 113px;
    box-shadow:  2px 2px 8px darken($light-medium-gray, 13%);
    padding: 6px 13px;
    display: inline-block;
    text-align: center;
    color: $medium-gray;
    margin: 4px calc(3px + 0.7vw);
    &:hover{
      background: lighten($green, 24%);
    }
    &.active{
      background: $green;
    }
    &.articleform__buttons--small{
      width: 45px;
      border: white;
    }
    &.articleform__buttons--publish{
      background: $bold-blue;
      color: white;
      &:hover{
        opacity:0.9;
      }
    }
    @media screen and ( min-width: 600px ){
      font-size:16px;
      width: 130px;
    }
  }
}

.articleform__overlay{
  position: absolute;
  top: 0px;
  left: 0;
  right: 0;
  bottom: 0;
  background: $dark-gray;
  color: white;
  padding: 5%;
  box-shadow: $bold-shadow;
  text-align: center;
  border-radius: 3px;
  h2{
    color: $green;
    font-size: 2em;
    font-family: $helvetica-condensed;
  }
  input[type="file"]{
    width: 90%;
    background: $black;
    padding: 20px;
    border-radius: 3px;
    margin: 20px auto;
    display: block;
    cursor:pointer;
  }
  input[type="text"]{
    width: calc(80% - 20px);
    padding: 10px 20px;
    border-radius: 3px;
    margin: 5px auto 0px;
    display: inline-block;
    border: 0px;
    font-size: 1em;
    @media screen and ( min-width: 600px ){
      width: calc(100% - 270px);
      border-top-left-radius: 0px;
      border-bottom-left-radius: 0px;
      margin: 20px auto 0px;
    }
  }
  label {
    margin-top:20px;
    display: inline-block;
    width: 80%;
    font-weight: bold;
    color: white;
    font-size: 1em;
    background: $medium-gray;
    padding: 10px;
    border-radius: 3px;
    @media screen and ( min-width: 600px ){
      width: 180px;
      border-top-right-radius: 0px;
      border-bottom-right-radius: 0px;  
    }

  }
  small {
    display: block;
    font-size:0.8em;
    color: $light-medium-gray;
    text-align: left;
    width: 80%;
    margin: 7px auto;
    font-style:oblique;
    @media screen and ( min-width: 600px ){
      width: calc(100% - 40px);
    }
    button {
      margin-left: 5px;
      background: $bold-blue;
    }
  }
  .articleform__exitbutton{
    position:absolute;
    color: white;
    font-size: 37px;
    font-weight: bold;
    left: 1%;
    top: 0%;
    background: transparent;
    border: 0px;
    cursor:pointer;
  }
  button{
    background: $red;
    color: white;
    border: 0px;
    padding: 4px 10px;
    font-size: 14px;
    border-radius: 3px;
    font-weight: bold;
  }
  .articleform__donebutton{
    display: block;
    font-size: 22px;
    background: $green;
    color: $black;
    padding: 10px 40px;
    margin: 20px auto;

  }
  img{
    display: block;
    margin: 10px auto;
    max-width: 80%;
  }
}

.articleform__notice{
  position: fixed;
  bottom:0;
  left:0;
  right:0;
  padding: 25px 0px;
  color: white;
  background: $green;
  text-align: center;
  font-size:2em;
  font-weight: bold;
  z-index:50;
  &.articleform__notice--draft{
    background: $yellow;
    color: $black;
  }
}