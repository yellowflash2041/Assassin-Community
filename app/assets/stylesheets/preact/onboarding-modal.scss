@import 'variables';
@import 'mixins';

.pages-onboarding {
  min-height: initial;
}

.onboarding-body {
  @include themeable(background-color, theme-background, $white);
  position: fixed;
  margin-top: 50px;
  bottom: 0;
  height: 100vh;
  width: 100%;
  overflow: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
  line-height: 200%;
  z-index: 11;
  margin: 0;
  background-image: url('/assets/onboarding-background-white.png');
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
  /* Center the image */
  background-position: center;
  /* Do not repeat the image */
  background-repeat: no-repeat;

  a {
    @include themeable(color, theme-anchor-color, #2e3b9e);
  }

  .sticker-logo {
    height: 80px;
    width: 80px;
    transform: rotate(-5deg);
    padding: 10px;
    vertical-align: middle;
    top: -10px;
    padding-top: 5px;
  }

  .onboarding-main {
    display: flex;
    flex-direction: column;
    width: 800px;
    height: 466px;
    @include themeable(background-color, theme-container-background, $white);
    border: 8px solid #6b55c9;
    @include themeable(border-color, theme-container-border, #6b55c9);
    padding: 30px;
    border-radius: 3px;
    z-index: 1;
    font-size: 20px;
    box-shadow: 0 0 8px 2px rgba(0, 0, 0, 0.3), 8px 0px 13px rgba(0, 0, 0, 0.21);
    align-items: center;
    justify-content: center;
  }

  .onboarding-content {
    flex: 1 0 auto;
    width: 100%;
    height: 440px;
    margin-top: 5px;

    h1 {
      margin-top: 15px;
      line-height: 1.1em;
    }
  }

  .onboarding-navigation {
    align-self: flex-end;
    flex-shrink: 0;
    width: 100%;
  }

  .background-image {
    z-index: -1;
    position: absolute;
    top: 60px;
    left: 0px;
  }

  h1 {
    text-align: center;
  }

  h2 {
    font-size: 15px;

    @media screen and (min-width: 600px) {
      font-size: 26px;
    }
  }

  p {
    line-height: 1.25em;
  }

  button {
    width: 130px;
    padding: 6px 0px;
    height: auto;
    margin: 11px 20px;
    text-align: center;
    font-weight: bold;
    border-radius: 3px;
    border: 2px solid #0a0a0a;
    @include themeable(border-color, theme-color, #0a0a0a);
    @include themeable(color, theme-color, #0a0a0a);
    @include themeable(background-color, theme-container-accent-background, $light-gray);
    font-family: 'HelveticaNeue-CondensedBold', 'HelveticaNeueBoldCondensed',
      'HelveticaNeue-Bold-Condensed', 'Helvetica Neue Bold Condensed',
      'HelveticaNeueBold', 'HelveticaNeue-Bold', 'Helvetica Neue Bold',
      'HelveticaNeue', 'Helvetica Neue', 'TeXGyreHerosCnBold', 'Helvetica',
      'Tahoma', 'Geneva', 'Arial Narrow', 'Arial', sans-serif;
    -webkit-appearance: none;
    font-stretch: condensed;
    font-size: 26px;
  }

  .next-button {
    @include themeable(background, theme-top-bar-write-background, #66e2d5);
    float: right;
  }

  .about {
    width: 100%;

    label,
    input,
    textarea {
      display: block;
      width: 95%;
      margin: 0 auto;
    }

    input,
    textarea {
      padding: 12px;
      font-size: 19px;
      text-emphasis: bold;
      border: 1px solid #dee6e9;
      border-radius: 3px;
    }

    textarea {
      height: 120px;
    }
  }

  .checkbox-slide {
    label {
      display: block;
      cursor: pointer;
    }

    input[type='checkbox'] {
      -webkit-appearance: none;
      width: 19px;
      height: 19px;
      background: transparent;
      border-radius: 5px;
      border: 2px solid #555;
      vertical-align: middle;
      margin: 8px;
      cursor: pointer;
    }

    input[type='checkbox']:checked {
      @include themeable(background, theme-anchor-color, #66e2d5);
    }
  }

  .warning-message {
    background-color: red;
    opacity: 0.7;
    padding: 10px;
    width: 100%;
    border-radius: 3px;
    color: white;
  }

  .scroll {
    height: 365px;
    overflow: auto;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    border-bottom: 1px solid $light-medium-gray;
  }

  .tag {
    padding: 44px 6px 46px;
    font-size: 20px;
    margin: 5px;
    border-radius: 3px;
    font-weight: bold;
    border: none;
    outline: none;
    overflow: hidden;
    position: relative;
    text-overflow: ellipsis;
    width: 100%;
    opacity: 0.8;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto,
      Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji',
      'Segoe UI Symbol';

    @media screen and (min-width: 600px) {
      width: 255px;
    }

    &.tag-selected {
      opacity: 1;
    }

    &:hover {
      opacity: 1;
    }

    .onboarding-tag-follow-indicator {
      position: absolute;
      bottom: 22px;
      left: 0px;
      right: 0px;
      text-align: center;
      font-size: 0.66em;
    }
  }

  .select-all-button-wrapper {
    width: 100%;

    button {
      width: 300px;
      line-height: 1em;
      text-align: left;
      padding-left: 15px;
    }
  }

  .user {
    width: calc(100% - 40px);
    overflow: hidden;
    text-overflow: ellipsis;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto,
      Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji',
      'Segoe UI Symbol';
    font-size: 15px;
    padding: 5px;
    border-radius: 3px;
    text-align: center;
    margin-right: 5px;
    position: relative;
    color: $black;

    .onboarding-user-follow-status {
      position: absolute;
      color: #32a99c;
      right: 8px;
      top: 5px;
      font-size: 0.88em;
    }

    span {
      display: block;
      width: 100%;
    }

    p {
      font-weight: 400;
      height: 55px;
    }

    @media screen and (min-width: 600px) {
      width: 365px;
    }

    img {
      height: 80px;
      width: 80px;
      border-radius: 50%;
      padding: 5px;
      margin: 10px auto;
      display: block;
    }
  }

  .user:active {
    background-color: red;
  }

  .badge-image {
    position: absolute;
  }

  .onboarding-what-next {
    display: flex;
    flex-direction: column;
    height: 250px;
    overflow: auto;

    @media screen and (min-width: 600px) {
      flex-direction: row;
    }

    a {
      background: #d9e7ff;
      padding: 15px 4px 0px 12px;
      height: 140px;
      margin: 4px auto;
      border-radius: 3px;
      width: 350px;
      display: block;
      line-height: 1.2em;
      font-weight: bold;
      position: relative;
      text-align: center;

      @media screen and (min-width: 600px) {
        height: 220px;
        margin: 4px 4px;
        text-align: left;
        padding: 20px 4px 0px 12px;
      }

      .whatnext-emoji {
        display: none;
        text-align: center;
        position: absolute;
        bottom: 22px;
        left: 0;
        right: 0;
        font-size: 50px;

        @media screen and (min-width: 600px) {
          display: block;
        }
      }
    }
  }

  .onboarding-previous-location {
    text-align: center;
    margin-top: 10px;
    font-weight: bold;
    display: none;

    @media screen and (min-width: 600px) {
      display: block;
    }

    code {
      font-size: 0.6em;
      margin-top: -10px;
      display: block;
      font-weight: 400;
    }
  }

  @media screen and (max-width: 800px) {
    .onboarding-main {
      width: 100%;
      height: 100%;
      background-image: url('/assets/mobile-onboarding-background.png');
      background-size: cover;
      /* Center the image */
      background-position: center;
      /* Do not repeat the image */
      background-repeat: no-repeat;
      border: none;
      color: $black;
    }

    .onboarding-content {
      flex: 0 0 auto;

      h1,
      h2 {
        margin-top: initial;
      }
    }

    span {
      display: block;
    }
  }

  @media screen and (max-width: 650px) {
    .onboarding-main {
      font-size: 18px;
    }

    button {
      width: 100px;
      font-size: 20px;
    }
  }

  @media screen and (max-width: 450px) {
    .onboarding-main {
      font-size: 16px;
    }

    button {
      width: 100px;
      font-size: 18px;
    }
  }

  @media screen and (max-width: 330px) {
    button {
      width: 80px;
      font-size: 14px;
    }

    .sticker-logo {
      height: 40px;
      width: 40px;
    }

    h1 {
      font-size: 20px;
    }
  }
}