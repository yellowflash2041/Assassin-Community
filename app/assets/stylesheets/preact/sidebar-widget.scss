@import 'variables';

.sidebar-additional{
  .widget{
    .widget-suggested-follows-container{
      h4{
        margin: 0;
      }
      .widget-body{
        .widget-list-item__suggestions{
          .widget-list-item__close-button{
            float: right;
            width: 24px;
            height: 19px;
            margin: 3px;
          }
          .widget-list-item__content{
            max-width: 180px;
            margin-left: 60px;
            margin-right: 20px;
            min-height: 55px;
            a{
              display: block;
              padding: 5px 0px;
              font-size: 18px;
              color: $black;
              color: var(--theme-color, $black);
              &:hover {
                text-decoration: underline;
              }
            }
            img{
              width: 48px;
              height: 48px;
              border-radius: 50%;
              position: absolute;
              top: unset;
              left: 15px;
            }
            .widget-list-item__follow-button{
              border: 2px solid transparent;
              border-radius: 3px;
              padding: 2px 8px;
              text-align: center;
              background-color: $green;
              color: $black;
              font-family: $helvetica-condensed;
              font-weight: 900;
              font-size: 0.8em;
            }
          }
        }
      }
    }
  }
}
