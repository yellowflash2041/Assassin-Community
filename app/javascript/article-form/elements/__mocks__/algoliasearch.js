global.document.head.innerHTML =
  "<meta name='algolia-public-id' content='abc123' />" +
  "<meta name='algolia-public-key' content='abc123' />" +
  "<meta name='environment' content='test' />";

const mockIndex = {
  search: (query, options) =>
    new Promise((resolve, reject) => {
      process.nextTick(() => {
        const searchResults = {
          gi: {
            hits: [
              {
                name: 'git',
                bg_color_hex: '#888751',
                text_color_hex: '#56c938',
                hotness_score: 0,
                supported: true,
                objectID: '4',
                _highlightResult: {
                  name: {
                    value: '<em>gi</em>t',
                    matchLevel: 'full',
                    fullyHighlighted: false,
                    matchedWords: ['gi'],
                  },
                  bg_color_hex: {
                    value: '#888751',
                    matchLevel: 'none',
                    matchedWords: [],
                  },
                  text_color_hex: {
                    value: '#56c938',
                    matchLevel: 'none',
                    matchedWords: [],
                  },
                },
              },
            ],
            nbHits: 1,
            page: 0,
            nbPages: 1,
            hitsPerPage: 10,
            processingTimeMS: 1,
            exhaustiveNbHits: true,
            query: 'gi',
            params:
              'query=gi&hitsPerPage=10&filters=supported%3Atrue&restrictIndices=searchables_development%2CTag_development%2Cordered_articles_development%2Cordered_articles_by_published_at_development%2Cordered_articles_by_positive_reactions_count_development%2Cordered_comments_development',
          },
        };

        const results = searchResults[query] || { hits: [] };

        resolve(results);
      });
    }),
};
const client = {
  initIndex: index => mockIndex,
};

export default jest.fn().mockImplementation((id, key) => client);
