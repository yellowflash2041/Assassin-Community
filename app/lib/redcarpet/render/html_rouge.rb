require "rouge/plugins/redcarpet"

module Redcarpet
  module Render
    class HTMLRouge < HTML
      include Rouge::Plugins::Redcarpet

      def link(link, _title, content)
        # Probably not the best fix but it does it's job of preventing
        # a nested links.
        return nil if /<a\s.+\/a>/.match?(content)

        link_attributes = ""
        @options[:link_attributes]&.each do |attribute, value|
          link_attributes += %( #{attribute}="#{value}")
        end
        %(<a href="#{link}"#{link_attributes}>#{content}</a>)
      end

      def header(title, header_number)
        anchor_link = slugify(title)
        <<~HEREDOC
          <h#{header_number}>
            <a name="#{anchor_link}" href="##{anchor_link}" class="anchor">
            </a>
            #{title}
          </h#{header_number}>
        HEREDOC
      end

      private

      def slugify(string)
        stripped_string = ActionView::Base.full_sanitizer.sanitize string
        stripped_string.downcase.gsub(EmojiRegex::Regex, "").strip.gsub(/[[:punct:]]/u, "").gsub(/\s+/, "-")
      end
    end
  end
end
