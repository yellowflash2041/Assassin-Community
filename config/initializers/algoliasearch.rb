AlgoliaSearch.configuration = {
  application_id: ApplicationConfig["ALGOLIASEARCH_APPLICATION_ID"],
  api_key: ApplicationConfig["ALGOLIASEARCH_API_KEY"]
}

if Rails.env.test?
  ::ALGOLIASEARCH_PUBLIC_SEARCH_ONLY_KEY = "test_test_test".freeze
else
  # Restrict Access to private indices
  params = {
    restrictIndices: [
      "searchables_#{Rails.env}",
      "Tag_#{Rails.env}",
      "ordered_articles_#{Rails.env}",
      "ordered_articles_by_published_at_#{Rails.env}",
      "ordered_articles_by_positive_reactions_count_#{Rails.env}",
      "ordered_comments_#{Rails.env}",
    ].join(",")
  }
  secured_algolia_key = Algolia.generate_secured_api_key(
    ApplicationConfig["ALGOLIASEARCH_SEARCH_ONLY_KEY"], params
  )

  ::ALGOLIASEARCH_PUBLIC_SEARCH_ONLY_KEY = secured_algolia_key
end
