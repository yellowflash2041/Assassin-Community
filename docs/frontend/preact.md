---
title: Preact
---

# Preact

[Preact](https://preactjs.com/) is an alternative to React with the same modern API.

Preact components are packaged using [Webpacker](/frontend/webpacker) and the Preact code is located in `app/javascript`.

The components are mounted when needed, look for `javascript_pack_tag` in the view pages inside `app/views`.
