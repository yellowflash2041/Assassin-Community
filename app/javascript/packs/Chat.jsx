import { h, render } from 'preact';
import Chat from '../chat/chat';

function loadElement() {
  const root = document.getElementById('chat');

  if (root) {
    render(<Chat {...root.dataset} />, root, root.firstChild);
    const placeholder = document.getElementById('chat_placeholder');
    if (placeholder) {
      root.removeChild(placeholder);
    }
  }
}

window.InstantClick.on('change', () => {
  loadElement();
});

loadElement();
