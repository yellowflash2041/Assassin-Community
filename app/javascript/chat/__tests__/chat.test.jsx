import { h } from 'preact';
import render from 'preact-render-to-json';
import { shallow } from 'preact-render-spy';
import fetch from 'jest-fetch-mock';
import { JSDOM } from 'jsdom';
import Chat from '../chat';
import algoliasearch from '../__mocks__/algoliasearchUsers';

global.fetch = fetch;

const doc = new JSDOM('<!doctype html><html><body></body></html>');
global.document = doc;
global.document.body.innerHTML =
  '<div class="chat chat--expanded" data-no-instant={true}><div class="chat__activechat"><div class="activechatchannel"><div class="activechatchannel__conversation"><div class="activechatchannel__header"></div><div class="activechatchannel__messages" id="messagelist"><div class="messagelist__sentinel" id="messagelist__sentinel"/></div><div class="activechatchannel__alerts"><div class="chatalert__default chatalert__default--hidden">More new messages below</div></div><div class="activechatchannel__form"><div class="messagecomposer"><textarea class="messagecomposer__input" id="messageform" maxLength="1000" onKeyDown={[Function]} placeholder="Message goes here" /><button class="messagecomposer__submit" onClick={[Function]}>SEND</button></div></div></div></div></div></div>';
global.window = doc.defaultView;
global.window.algoliasearch = algoliasearch;

// mock observer and user ID
window.IntersectionObserver = jest.fn(function() {
  this.observe = jest.fn();
});
global.window.currentUser = { id: 'some_id' };

// fake props to pass to Chat
const rootData = {
  algoliaId: 'somealgoliaid',
  algoliaIndex: 'somealgoliaindex',
  algoliaKey: 'somealgoliakey',
  chatChannels: JSON.stringify([
    {
      channel_name: 'channel name 1',
      last_opened_at: 'September 2, 2018',
      channel_users: [],
      last_message_at: 'September 21, 2018',
      channel_type: 'group',
      slug: '0',
      id: 12345,
      messages_count: 124,
    },
    {
      channel_name: 'group channel 2',
      last_opened_at: 'September 12, 2018',
      channel_users: [
        {
          profile_image: 'fake_profile_image',
          darker_color: '#111111',
          last_opened_at: 'some last open date',
        },
        {
          profile_image: 'fake_profile_pic',
          darker_color: '#222222',
          last_opened_at: 'some other last open date',
        },
      ],
      last_message_at: 'September 14, 2018',
      channel_type: 'direct',
      slug: '1',
      id: 34561,
      messages_count: 83,
    },
    {
      channel_name: 'group channel 3',
      last_opened_at: 'September 30, 2018',
      channel_users: [
        {
          profile_image: 'fake_profile_image',
          darker_color: '#111111',
          last_opened_at: 'some last open date',
        },
        {
          profile_image: 'fake_profile_pic',
          darker_color: '#222222',
          last_opened_at: 'some other last open date',
        },
      ],
      last_message_at: 'September 29, 2018',
      channel_type: 'group',
      slug: '2',
      id: 67890,
      messages_count: 56,
    },
  ]),
  chatOptions: JSON.stringify({
    showChannelsList: true,
    showTimestamp: true,
    activeChannelId: 34561,
  }),
  githubToken: 'somegithubtoken',
  pusherKey: 'somepusherkey',
};

const getChat = () => <Chat {...rootData} />;

describe('<Chat />', () => {
  it('should load chat', () => {
    const tree = render(getChat());
    expect(tree).toMatchSnapshot();
  });

  it('should have the proper elements, attributes and content', () => {
    const context = shallow(getChat());
    expect(context.exists()).toEqual(true);

    expect(context.find('.chat').exists()).toEqual(true);
    expect(context.find('.chat--expanded').exists()).toEqual(true);
    expect(context.find('.chat__activechat').exists()).toEqual(true);

    // renderChatChannels
    expect(context.find('.chat__channels').exists()).toEqual(true);
    expect(context.find('.chat__channelstogglebutt').exists()).toEqual(true);
    expect(context.find('.chat__channeltypefilter').exists()).toEqual(true);
    expect(context.find('.chat__channeltypefilter').text()).toEqual(
      'alldirectgroup',
    );

    // renderActiveChatChannel
    expect(context.find('.activechatchannel').exists()).toEqual(true);
    expect(context.find('.activechatchannel__conversation').exists()).toEqual(
      true,
    );
    expect(context.find('.activechatchannel__header').exists()).toEqual(true);
    expect(context.find('.activechatchannel__messages').exists()).toEqual(true);
    expect(context.find('.activechatchannel__alerts').exists()).toEqual(true); // div that wraps Alert
    expect(context.find('.activechatchannel__form').exists()).toEqual(true); // div that wraps Compose
  });

  it('should un-expand and expand chat channels properly', () => {
    const context = shallow(getChat());

    // un-expand chat channels
    context.find('.chat__channelstogglebutt').simulate('click');
    expect(context.find('.chat__channeltypefilter').exists()).toEqual(false); // now hidden
    expect(context.find('.chat__channeltypefilter').text()).toEqual(''); // no text

    // re-expand chat channel
    context.find('.chat__channelstogglebutt').simulate('click');
    expect(context.find('.chat__channeltypefilter').exists()).toEqual(true);
    expect(context.find('.chat__channeltypefilter').text()).toEqual(
      'alldirectgroup',
    );
  });
});
