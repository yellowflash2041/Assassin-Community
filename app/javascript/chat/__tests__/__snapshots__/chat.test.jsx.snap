// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> should load chat 1`] = `
<div
  class="chat chat--expanded"
  data-no-instant={true}
>
  <div
    class="chat__channels chat__channels--expanded"
  >
    <button
      class="chat__channelstogglebutt"
      onClick={[Function]}
      type="button"
    >
      &lt;
    </button>
    <input
      onKeyUp={[Function]}
      placeholder="Filter"
    />
    <div
      class="chat__channeltypefilter"
    >
      <button
        class="chat__channeltypefilterbutton chat__channeltypefilterbutton--active"
        data-channel-type="all"
        onClick={[Function]}
        type="button"
      >
        all
      </button>
      <button
        class="chat__channeltypefilterbutton chat__channeltypefilterbutton--inactive"
        data-channel-type="direct"
        onClick={[Function]}
        type="button"
      >
        direct
      </button>
      <button
        class="chat__channeltypefilterbutton chat__channeltypefilterbutton--inactive"
        data-channel-type="invite_only"
        onClick={[Function]}
        type="button"
      >
        group
      </button>
    </div>
    <div
      class="chatchannels"
    >
      <div
        class="chatchannels__channelslist"
        id="chatchannels__channelslist"
      >
        <button
          class="chatchanneltabbutton"
          onClick={[Function]}
          type="button"
        >
          <span
            class="chatchanneltab chatchanneltab--inactive chatchanneltab--old"
            style={
              Object {
                "border": "1px solid undefined",
                "boxShadow": "3px 3px 0px undefined",
              }
            }
          >
            <span
              class="chatchanneltabindicator chatchanneltabindicator--old"
            >
              <img
                alt="pic"
                class="chatchanneltabindicatordirectimage"
              />
            </span>
            channel name 1
          </span>
        </button>
        <button
          class="chatchanneltabbutton"
          onClick={[Function]}
          type="button"
        >
          <span
            class="chatchanneltab chatchanneltab--inactive chatchanneltab--old"
            style={
              Object {
                "border": "1px solid undefined",
                "boxShadow": "3px 3px 0px undefined",
              }
            }
          >
            <span
              class="chatchanneltabindicator chatchanneltabindicator--old"
            >
              <img
                alt="pic"
                class="chatchanneltabindicatordirectimage"
              />
            </span>
            group channel 2
          </span>
        </button>
        <button
          class="chatchanneltabbutton"
          onClick={[Function]}
          type="button"
        >
          <span
            class="chatchanneltab chatchanneltab--inactive chatchanneltab--old"
            style={
              Object {
                "border": "1px solid undefined",
                "boxShadow": "3px 3px 0px undefined",
              }
            }
          >
            <span
              class="chatchanneltabindicator chatchanneltabindicator--old"
            >
              <img
                alt="pic"
                class="chatchanneltabindicatordirectimage"
              />
            </span>
            group channel 3
          </span>
        </button>
      </div>
      <div
        class="chatchannels__config"
      >
        <img
          alt=""
          src=""
          style={
            Object {
              "height": "18px",
            }
          }
        />
        <div
          class="chatchannels__configmenu"
        >
          <a
            href="/settings"
          >
            DEV Settings
          </a>
          <a
            href="/report-abuse"
          >
            Report Abuse
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="chat__activechat"
  >
    <div
      class="activechatchannel"
    >
      <div
        class="activechatchannel__conversation"
      >
        <div
          class="activechatchannel__header"
        >
          Â 
        </div>
        <div
          class="activechatchannel__messages"
          id="messagelist"
        >
          <div
            class="messagelist__sentinel"
            id="messagelist__sentinel"
          />
        </div>
        <div
          class="activechatchannel__alerts"
        >
          <div
            class="chatalert__default chatalert__default--hidden"
          >
            More new messages below
          </div>
        </div>
        <div
          class="activechatchannel__form"
        >
          <div
            class="messagecomposer"
          >
            <textarea
              class="messagecomposer__input"
              id="messageform"
              maxLength="1000"
              onKeyDown={[Function]}
              placeholder="Message goes here"
            />
            <button
              class="messagecomposer__submit"
              onClick={[Function]}
            >
              SEND
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
